main/src/ca/uwaterloo/flix/language/ast/TypedAst.scala

```diff
@@ -120,7 +121,11 @@ object TypedAst {
       def eff: Type = Type.Pure
     }
 
-    case class ApplyClo(exp1: Expr, exp2: Expr, tpe: Type, eff: Type, loc: SourceLocation) extends Expr
+    /** `exps` and `evars` have the same, non-zero length. */
+    case class ApplyClo(exp: Expr, exps: List[Expr], argEffs: List[Type], tpe: Type, eff: Type, loc: SourceLocation) extends Expr {
+      if (exps.sizeIs != argEffs.size) throw InternalCompilerException("Inconsistent apply closure arity", loc)
+      else if (exps.isEmpty) throw InternalCompilerException("Unexpected empty closure apply", loc)
+    }
```

Comment:
This design is not the nicest but with the active checking I think it is acceptable, what do you think @mlutze ?

Reply:
We could have `Nel[(Expr, Type)]` - do we have Nel in scala? couldn't find it

Reply:
Do we actually need the effs list?

Scala has no `Nel`. You could use `::` as a type to achieve it, but mapping over a `::` returns a `List`. There's a discussion somewhere where Martin Odersky says he's against `Nel` in Scala "because what about `TwoList` etc.?" But clearly we would never have a need to go that far.

Reply:
i've commented below, but i need them to reconstruct the individual applications for the backend

classic argument with the lists - everything i slippery